dokan/loaddll.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2018 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+// +build windows
+
+package dokan
+
+/*
+#include "bridge.h"
+*/
+import "C"
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"os"
+	"runtime"
+	"strings"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
+const shortPath = `DOKAN1.DLL`
+const syswow64 = `C:\WINDOWS\SYSWOW64\`
+const system32 = `C:\WINDOWS\SYSTEM32\`
+
+type errorPrinter struct {
+	buf bytes.Buffer
+}
+
+func (ep *errorPrinter) Printf(s string, os ...interface{}) {
+	fmt.Fprintf(&ep.buf, s, os...)
+}
+
+// loadLibrary calls win32 LoadLibrary and logs the result.
+func loadLibrary(epc *errorPrinter, path string) (windows.Handle, error) {
+	hdl, err := windows.LoadLibrary(path)
+	epc.Printf("LoadLibrary(%q) -> %v,%v\n", path, hdl, err)
+	return hdl, err
+}
+
+// doLoadDLL tries to load the dokan DLL from various locations.
+func doLoadDLL(epc *errorPrinter, path string) (windows.Handle, error) {
+	var guessPath bool
+	epc.Printf("loadDokanDLL %q\n", path)
+	if path == "" {
+		path = shortPath
+		guessPath = true
+	} else {
+		path = strings.Replace(path, `/`, `\`, -1)
+	}
+	const loadLibrarySearchSystem32 = 0x800
+	const flags = loadLibrarySearchSystem32
+	hdl, err := windows.LoadLibraryEx(path, 0, flags)
+	epc.Printf("loadDokanDLL LoadLibraryEx(%q,0,%x) -> %v,%v\n", path, flags, hdl, err)
+	if err == nil || !guessPath {
+		return hdl, err
+	}
+	// User probably has not installed KB2533623 which is a security update
+	// from 2011. Without this Windows security update loading libraries
+	// is unsafe on Windows.
+	// Continue to try to load the DLL regardless.
+
+	if runtime.GOARCH == `386` {
+		hdl, err = loadLibrary(epc, syswow64+shortPath)
+		if err == nil {
+			return hdl, nil
+		}
+	}
+	hdl, err = loadLibrary(epc, system32+shortPath)
+	if err == nil {
+		return hdl, nil
+	}
+	hdl, err = loadLibrary(epc, shortPath)
+	if err == nil {
+		return hdl, nil
+	}
+	err = fmt.Errorf("loadDokanDLL: cannot load Dokan DLL: %v", err)
```

Comment:
Would it be worthwhile to log each of the above non-nil errors along with which path caused them, in case each is different?

Reply:
The loadLibrary with the epc argument logs every call to the epc with the path included, and that gets attached to the error value in loadDokanDLL.

Reply:
Ah right.  So I guess it's not strictly needed to log `err` again here.